doc/doxygen/src/driver-guide.md

```diff
@@ -0,0 +1,351 @@
+anyone wanting to use the driver can find out the supported features without
+having to scan through the code.
+
+### Device descriptor and parameter configuration         {#driver-guide-types}
+Each device MUST supply a data structure, holding the devices state and
+configuration, using the naming scheme of `DEVNAME_t` (e.g. `dht_t`, or
+`at86rf2xx_t`). In context of RIOT, we call this data structure the device
+descriptor.
+
+This device descriptor MUST contain all the state data of a device. By this, we
+are not limited on the number of instances of the driver we can run
+concurrently. The descriptor is hereby used for identifying the device we want
+to interact with, and SHOULD always be the first parameter for all device driver
+related functions.
+
+Typical things found in these descriptors are e.g. used peripherals (e.g. SPI or
+I2C bus used, interfacing GPIO pins), data buffers (e.g. RX/TX buffers where
+needed), or state machine information.
+
+On top of the device descriptor, each device driver MUST also define a data
+structure holding the needed configuration data. The naming scheme for this type
+is `DEVNAME_params_t`. In contrary to the device descriptor, this data structure
+should only contain static information, that is needed for the device
+initialization as it is preferably allocated in ROM.
+
+A simple I2C temperature sensors's device descriptor could look like this:
+```c
+typedef struct {
+    tmpabc_params_t p;  /**< device configuation parameter like I2C bus and bus addr */
+    int scale;          /**< some custom scaling factor for converting the results */
+} tmpabc_t;
+
+/* with params being */
+typedef struct {
+    i2c_t bus;          /**< I2C bus the device is connected to */
+    uint8_t addr;       /**< the device's address on the bus */
+} tmpabc_params_t;
+```
+
+**NOTE:** In many cases it makes sense, to copy the `xxx_params` data into the
+device descriptor during initialization. In some cases, it is however better to
+just link the `params` data via pointer and only copy selected data. This way,
+configuration data that is only used once can be read directly from ROM, while
+often used fields (e.g. used peripherals) are stored directly in the device
+descriptor and one saves hereby one de-referencing step when accessing them.
+
+### Default device configuration                         {#driver-guide-config}
+Each device driver in RIOT MUST supply a default configuration file, named
+`DEVNAME_params.h`. This file should be located in the `RIOT/drivers/...`. The
+idea is, that this file can be overridden by an application or a board, by
+simply putting a file with the same name in the application's or the board's
+include folders, while RIOT's build system takes care of prefferring those files
+instead of the default params file.
+
+A default parameter header file for the example temperature sensor above would
+look like this (`tmpabc_params.h`):
+```c
```

Comment:
Doxygen markdown doesn't now this kind of code-block guards. Use

```markdown
~~~~~~~~~~~~~~~~ {.c}
/* code here */
~~~~~~~~~~~~~~~~
```

Reply:
It [actually does](http://doxygen.nl/manual/markdown.html#md_fenced) in the current stable version, but not in the version used in the CI. Thanks for spotting this!

Reply:
I think it actually also does on the CI version, but not with syntax highlighting...

Reply:
(and the `c` is throwing it off...)